src/multicall.rs

```diff
@@ -0,0 +1,133 @@
+use alloy::network::{ReceiptResponse, TransactionBuilder};
+use alloy::providers::Provider;
+use alloy::rpc::types::TransactionRequest;
+use alloy::{
+    providers::{Network, RootProvider},
+    transports::{Transport, TransportErrorKind, TransportResult},
+};
+use alloy_primitives::{Address, Bytes};
+use alloy_sol_types::{sol, SolCall};
+use futures::future::join_all;
+use oneshot::{Receiver, Sender};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use tokio::time::sleep;
+
+// Define the Multicall3 interface
+sol! {
+    interface Multicall3 {
+        struct Call3 {
+            address target;
+            bytes callData;
+            bool allow_failure;
+        }
+        struct Result {
+            bool success;
+            bytes returnData;
+        }
+        function aggregate3(Call3[] calldata calls) external returns (Result[] memory returnData);
+    }
+}
+
+struct BatchedCall {
+    target: Address,
+    call_data: Bytes,
+    allow_failure: bool,
+    callback: tokio::sync::oneshot::Sender<TransportResult<Bytes>>,
+}
+
+pub struct BatchingProvider<T, N> {
+    inner: Arc<RootProvider<T, N>>,
+    batch: Arc<Mutex<Vec<BatchedCall>>>,
```

Comment:
Why `Mutex`?

Reply:
had some ideas for a multi thread kind of stuff, but will scrape it